src/Analysers/LikelihoodRecoAnalyser.cpp

```diff
@@ -0,0 +1,180 @@
+
+		ParticlePointer TopHad = bestSolution->hadronicTop;
+		ParticlePointer TopLep = bestSolution->leptonicTop;
+		ParticlePointer W = bestSolution->hadronicW;
+		// double Nu = bestSolution->neutrinoChi2;
+
+		treeMan_->Fill("LeptonicTop_Pt",TopLep->pt() );
+		treeMan_->Fill("LeptonicTop_Eta",TopLep->eta() );
+		treeMan_->Fill("LeptonicTop_Energy",TopLep->energy() );
+		treeMan_->Fill("LeptonicTop_Mass",TopLep->mass() );
+
+		treeMan_->Fill("HadronicTop_Pt",TopHad->pt());
+		treeMan_->Fill("HadronicTop_Eta",TopHad->eta() );
+		treeMan_->Fill("HadronicTop_Energy",TopHad->energy());
+		treeMan_->Fill("HadronicTop_Mass",TopHad->mass() );
+
+		// Require one of the event selections to have been satisfied, and we have a genuine ttbar event
+		int selectionCriteria = -1;
+		if ( event->PassesElectronTriggerAndSelection() && event->isSemiLeptonicElectron() ) selectionCriteria = SelectionCriteria::ElectronPlusJetsReference;
+		else if ( event->PassesMuonTriggerAndSelection() && event->isSemiLeptonicMuon() ) selectionCriteria = SelectionCriteria::MuonPlusJetsReference;
+		if ( selectionCriteria < 0 ) {
+			TypeofSolution = 4;
+			treeMan_->Fill("TypeofSolution", TypeofSolution );
+		}
+
+
+		for ( unsigned int AllJetIndex=0; AllJetIndex < AllJets.size(); ++AllJetIndex ) {
+
+			JetPointer AllJet = AllJets[AllJetIndex];
+
+			if ( AllJet->matched_parton() == 0 ) continue;//if null pointer
+
+			if ( AllJet->ttbar_decay_parton() == 3 ){
+				// cout << "Found qbar decay parton" << endl;
+				++ttbarpartonsfound;
+			}
+			if ( AllJet->ttbar_decay_parton() == 4 ){
+				// cout << "Found q decay parton" << endl;
+				++ttbarpartonsfound;
+			}		
+			if ( AllJet->ttbar_decay_parton() == 5 ){
+				// cout << "Found leptonicb decay parton" << endl;
+				++ttbarpartonsfound;
+			}		
+			if ( AllJet->ttbar_decay_parton() == 6 ){
+				// cout << "Found hadronicb decay parton" << endl;
+				++ttbarpartonsfound;
+			}
+		}
+		// cout << "-------------------" << endl;
+		if ( ttbarpartonsfound != 4 ){
+			TypeofSolution = 3;
+			treeMan_->Fill("TypeofSolution", TypeofSolution);
+		}
+
+		double Discrim = bestSolution->discriminator;
+		double CSVDiscrim = bestSolution->CSVDiscriminator;
+		double MassDiscrim = bestSolution->MassDiscriminator;
+		double NuChi2Discrim = bestSolution->NuChi2Discriminator;
+
+		treeMan_->Fill( "likelihoodRatioDiscrimBest", Discrim );
+		treeMan_->Fill( "CSVDiscrimBest", CSVDiscrim );
+		treeMan_->Fill( "NuChi2DiscrimBest", NuChi2Discrim );
+		treeMan_->Fill( "MassDiscrimBest", MassDiscrim );
+
+
+
+		if (bestSolution->isCorrect()){
+			if (TypeofSolution == 0){
+				TypeofSolution = 1;
+				treeMan_->Fill("TypeofSolution", TypeofSolution);
+			}
+			
+			EventCorrect = 1;
+			treeMan_->Fill("EventReconstruction", EventCorrect);
+		}
+
+		else{
+			if (TypeofSolution == 0){
+				TypeofSolution = 2;
+				treeMan_->Fill("TypeofSolution", TypeofSolution);
+			}
+			treeMan_->Fill("EventReconstruction", EventCorrect);
+		}
+
```

Comment:
I would suggest only doing treeMan_->Fill("TypeofSolution", TypeofSolution) once per event, to make sure you only fill it once.  Set TypeofSolution as you work through all the different categories, and then fill the branch once.